### PR TITLE
Add plugins to rainbow bridge contracts

### PR DESCRIPTION
**Pull request by mfornet**
Use [near-plugins](https://github.com/mfornet/near-plugins/) to extend bridge contracts with new functionalities.
1. `Ownable` pattern allows setting the owner to a DAO that can control admin functions from the bridge.
2. `Pausable` changes the old mechanism for pausing functions, but provides the same features.
3. `Upgradable` allows owner to upgrade the bridge to a new version.
4. `FullAccessKeyFallback` allows the owner to attach a full access key in case of Emergency.

## Notes for reviewer: 

Here is the new code with all macros expanded. [Gist](https://gist.github.com/mfornet/bedc71fe0d0122540bb07718f76b713b)

The file is too large to be reviewed completely, so I recommend check the diff before and after the expansion and review it. Here are some examples of the new code ([1](https://gist.github.com/mfornet/bedc71fe0d0122540bb07718f76b713b#file-client-rs-L1743), [2](https://gist.github.com/mfornet/bedc71fe0d0122540bb07718f76b713b#file-client-rs-L282)).

This was generated using [cargo-expand](https://github.com/dtolnay/cargo-expand). It can be installed with `cargo expand`. To generate use the following command:

```bash
cd contracts/near/eth-client
cargo expand --target wasm32-unknown-unknown > generated.rs
```

# Testing plan
- All bridge tests should pass, i.e. basic mechanisms of the bridge haven't changed.
- WIP: Add tests to check all new features works as expected

### CODE REVIEW

File: contracts/near/eth-types/Cargo.toml [hunk 1]
```diff
@@ -9,8 +9,8 @@ crate-type = ["cdylib", "rlib"]
 
 [dependencies]
 serde = { version = "1.0", features = ["derive"] }
-near-sdk = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
-borsh = "0.8.1"
+near-sdk = "4.0.0-pre.5"
```
Comment (karim-en) [inline contracts/near/eth-types/Cargo.toml:12]:
```suggestion
near-sdk = "4.0.0-pre.7"
```

File: contracts/near/eth-types/Cargo.toml [hunk 2]
```diff
@@ -9,8 +9,8 @@ crate-type = ["cdylib", "rlib"]
...
+borsh = "0.9"
```
Comment (karim-en) [inline contracts/near/eth-types/Cargo.toml:13]:
```suggestion
borsh = "0.9.3"
```

File: contracts/near/rust-toolchain [hunk 3]
```diff
@@ -1 +1 @@
-stable-2020-10-08
+stable-2022-01-13
```
Comment (karim-en) [inline contracts/near/rust-toolchain:1]:
```suggestion
stable-2022-01-19
```

File: contracts/near/eth-prover/rust-toolchain [hunk 4]
```diff
@@ -1 +1 @@
-nightly-2020-10-08
+stable-2022-01-13
```
Comment (karim-en) [inline contracts/near/eth-prover/rust-toolchain:1]:
```suggestion
stable-2022-01-19
```

File: contracts/near/eth-prover/Cargo.toml [hunk 5]
```diff
@@ -10,13 +10,13 @@ crate-type = ["cdylib", "rlib"]
 [dependencies]
 wee_alloc = { version = "0.4.5", default-features = false, features = [] }
 eth-types =  { path = "../eth-types" }
-admin-controlled =  { path = "../admin-controlled" }
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
-near-sdk = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
-borsh = "0.8.1"
+near-sdk = "4.0.0-pre.5"
```
Comment (karim-en) [inline contracts/near/eth-prover/Cargo.toml:15]:
```suggestion
near-sdk = "4.0.0-pre.7"
```

File: contracts/near/eth-client/rust-toolchain [hunk 6]
```diff
@@ -1 +1 @@
-nightly-2020-10-08
+stable-2022-01-13
```
Comment (karim-en) [inline contracts/near/eth-client/rust-toolchain:1]:
```suggestion
stable-2022-01-19
```

File: contracts/near/eth-client/Cargo.toml [hunk 7]
```diff
@@ -9,9 +9,8 @@ crate-type = ["cdylib", "rlib"]
 
 [dependencies]
 eth-types =  { path = "../eth-types" }
-admin-controlled =  { path = "../admin-controlled" }
-near-sdk = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
-borsh = "0.8.1"
+near-sdk = "4.0.0-pre.5"
+borsh = "0.9"
```
Comment (karim-en) [inline contracts/near/eth-client/Cargo.toml:13]:
```suggestion
borsh = "0.9.3"
```

File: contracts/near/eth-client/Cargo.toml [hunk 8]
```diff
@@ -9,9 +9,8 @@ crate-type = ["cdylib", "rlib"]
 
 [dependencies]
 eth-types =  { path = "../eth-types" }
-admin-controlled =  { path = "../admin-controlled" }
-near-sdk = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
-borsh = "0.8.1"
+near-sdk = "4.0.0-pre.5"
```
Comment (karim-en) [inline contracts/near/eth-client/Cargo.toml:12]:
```suggestion
near-sdk = "4.0.0-pre.7"
```

File: contracts/near/admin-controlled/Cargo.toml [hunk 9]
```diff
@@ -8,4 +8,4 @@ edition = "2018"
 crate-type = ["cdylib", "rlib"]
 
 [dependencies]
-near-sdk = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
\ No newline at end of file
+near-sdk = "4.0.0-pre.5"
```
Comment (karim-en) [inline contracts/near/admin-controlled/Cargo.toml:11]:
```suggestion
near-sdk = "4.0.0-pre.7"
```

File: contracts/near/eth-prover/Cargo.toml [hunk 10]
```diff
@@ -10,13 +10,13 @@ crate-type = ["cdylib", "rlib"]
...
+borsh = "0.9"
```
Comment (karim-en) [inline contracts/near/eth-prover/Cargo.toml:16]:
```suggestion
borsh = "0.9.3"
```

**Review by karim-en — CHANGES_REQUESTED** [review]
Looks good to me, just proposed newer lib versions. 
Also please don't forget to update the binaries.

File: contracts/near/eth-client/rust-toolchain (hunk 6 above)
Comment (mfornet) [reply contracts/near/eth-client/rust-toolchain:1]: ❯ cargo build --target wasm32-unknown-unknown --release
info: syncing channel updates for 'stable-2022-01-19-x86_64-apple-darwin'
error: no release found for 'stable-2022-01-19'

**Review by karim-en — APPROVED** [review]

**Review by karim-en — APPROVED** [review]

**Review by sept-en — CHANGES_REQUESTED** [review]
Closed as it will be superseded by #845 with the newest version of near-plugins.

**Pull request closed without merging**